### DISCUSSION TITLE
文字列データ型の種類

### DISCUSSION THREAD

**Discussion by yuyakawai0411** [kakeru-one/web-engineering-roundtable#66] [General]
### 非バイナリ文字列のデータ型
| データ型 | 長さ | サイズ | 特徴 |
| --- | --- | --- | --- |
| CHAR型 | 固定長 | 0〜255文字 | データの実体を保存 |
| VARCHAR型 | 可定長 | 0〜65,535文字 | データの実体を保存 |
| TEXT型 | 可定長 | 0〜65,535byte以上(byteなのに注意) | データのポインタを保存 |

### バイナリ文字列のデータ型
| データ型 | 長さ | サイズ |  特徴 |
| --- | --- | --- | --- |
| BINARY型 |  固定長 | 0〜255byte | データの実体を保存 |
| VARBINARY型 | 可定長 | 0〜65,535byte | データの実体を保存 |
| BLOB型 | 可定長 | 0〜65,535byte以上 | データのポインタを保存 |

### リスト構造を持つデータ型
| データ型 | 特徴 |
| --- | --- |
| ENUM型 | 定義したリストの中から1つを選択する |
| SET型 | 定義したリストの中から複数を選択する |

## どのデータ型を選択すべきか？
- メモリ無駄に消費しないよう、データ型が小さい方を選ぶ
- 変換という無駄な処理を挟まないよう、扱う形式に近いデータ型を選ぶ
  - 例えば、文字として扱うのなら非バイナリ文字列のデータ型を使う。バイナリ形式で文字コードを使って変換するのは非効率のため

## 参考文献
https://dev.mysql.com/doc/refman/8.0/ja/string-types.html

**Comment by yuyakawai0411** [comment]
## 非バイナリ文字列とバイナリ文字列の違い
#### バイナリ文字列
人間が理解できない文字列。画像データやPDFなどのファイルを変換するときに使う。x〇〇は2桁の16進数を示す
```
"\x89PNG\r\n\x1A\n\x00\x00\x00\rIHDR\x00\x00\x00\x10\x00\x00\x00\x10"
```
#### 非バイナリ文字列
人間が読める文字列。文章などをそのまま保存するのに向く
```
'Hello world'
```

**Comment by kakeru-one** [comment]
TEXT型も65,535バイトまでしか保存できないので、保存する文字数だけでなく、1文字何バイトなのかを気にする必要がありそう

https://gihyo.jp/dev/serial/01/mysql-road-construction-news/0044#sec5:~:text=%E8%89%AF%E3%81%84%E3%81%A7%E3%81%97%E3%82%87%E3%81%86%E3%80%82-,BLOB%E5%9E%8B%E3%81%A8TEXT%E5%9E%8B,-BLOB%EF%BC%88Binary%20Large

**Reply by kakeru-one** [reply]
参考: https://tech.itandi.co.jp/entry/2022/08/26/180000

**Reply by kakeru-one** [reply]
> VARCHAR(65535)型は 文字数で数えて65535まで 記録する
> TEXT型は バイト数で数えて65535まで 記録する

**Comment by kakeru-one** [comment]
## TEXT型とVARCHAR型の保存のされ方による使い分け

> VARCHAR型は入力された文字列をテーブルの行内にそのまま保存するのに対し、TEXT型はデータ自体を別の場所に格納し、そのポインタのみをテーブルに保持します。

Reference: https://qiita.com/kondo0602/items/7c5f50986455feecc50f#text%E5%9E%8B

TEXT型とVARCHAR型では、レコード保存時の保存場所が違います。
VARCHAR型は行内にデータを保存するため、特定のケースではディスクI/Oが速くなることがあります。
TEXT型はデータを外部に保存するため、行サイズを小さくできるが、追加のI/Oが発生することがあります。

また、VARCHAR型は最大65,535文字まで保存できるが、
TEXT型は65,535バイトまで保存できます。このため、保存する文字数だけでなく、エンコーディングによる1文字あたりのバイト数を考慮する必要があります。


## MySQLにおける行サイズの制限について
Reference: https://dev.mysql.com/doc/refman/8.4/en/column-count-limit.html

以下のように書いてあります。

> The internal representation of a MySQL table has a maximum row size limit of 65,535 bytes, even if the storage engine is capable of supporting larger rows. [BLOB](https://dev.mysql.com/doc/refman/8.4/en/blob.html) and [TEXT](https://dev.mysql.com/doc/refman/8.4/en/blob.html) columns only contribute 9 to 12 bytes toward the row size limit because their contents are stored separately from the rest of the row.

**Reply by kakeru-one** [reply]
https://qiita.com/kondo0602/items/7c5f50986455feecc50f#text%E5%9E%8B

**Reply by kakeru-one** [reply]
https://dev.mysql.com/doc/refman/8.0/ja/table-size-limit.html

**Reply by yuyakawai0411** [reply]
TEXT型とBLOB型はデータの実態を持たず、データを格納しているメモリ番号(ポインタ)を持つ。
それゆえに、TEXT型とBLOB型は9〜12バイトしか使わない。
MySQLには行ごとに持てるbyte数の制約があるため、そのbyte数を削減する意味では、TEXTおよびBLOB型を使う価値がある

https://dev.mysql.com/doc/refman/8.0/ja/column-count-limit.html

**Reply by kakeru-one** [reply]
インデックスの最大キー長も理解しておく必要があり、
基本的にデフォルトだと767バイトが最大キー長となっているので、
場合によっては全てをDBに保存するのではなく、digestをDBに保存することで、先頭767バイト以内をindexを用いた検索に利用することができたりする。
（特定の行形式（DYNAMICまたはCOMPRESSED）を使用する場合、最大キー長が3072バイトになる。）
この時、本体の大きな文字列は、別テーブルもしくは、NoSQLに保存するのが良いと考えられる。

Reference①: https://gihyo.jp/dev/serial/01/mysql-road-construction-news/0032
Reference②: https://dev.mysql.com/doc/refman/8.0/ja/innodb-limits.html
> [DYNAMIC](https://dev.mysql.com/doc/refman/8.0/ja/glossary.html#glos_dynamic_row_format) または [COMPRESSED](https://dev.mysql.com/doc/refman/8.0/ja/glossary.html#glos_compressed_row_format) の行形式を使用する InnoDB テーブルでは、インデックスキーの接頭辞の長さの制限は 3072 バイトです。
> [REDUNDANT](https://dev.mysql.com/doc/refman/8.0/ja/glossary.html#glos_redundant_row_format) または [COMPACT](https://dev.mysql.com/doc/refman/8.0/ja/glossary.html#glos_compact_row_format) の行形式を使用する InnoDB テーブルのインデックスキー接頭辞の長さ制限は 767 バイトです。 たとえば、utf8mb4 文字セットおよび各文字の最大 4 バイトを想定して、TEXT または VARCHAR カラムで 191 文字を超える [column prefix](https://dev.mysql.com/doc/refman/8.0/ja/glossary.html#glos_column_prefix) インデックスを使用して、この制限に達する場合があります。
